### PR TITLE
adds authrep function with OAuth method

### PR DESCRIPTION
**Pull request by avilatusell**
Adds Authrep to authorize an application and report the associated transaction at the same time in the OAuth authentication pattern.
TO DO: tests

### CODE REVIEW

File: src/client.coffee [hunk 1]
```diff
@@ -238,6 +238,45 @@ module.exports = class Client
...
+         sys.puts "#{response.error_message} with code: #{response.error_code}"
+  ###
+
+  authrep_with_oauth: (options, callback) ->
+    this._verify_app_id_exists options
+
+    this._ensure_hits_exist_in options
+
+    query = this._build_query_string options
+    req_opts = this._prepare_request "/transactions/authrep.xml", query
```
Comment (kevprice83) [inline src/client.coffee]: the request should be `/transactions/oauth_authrep.xml` as you can see documented [here](https://github.com/3scale/nginx-oauth-templates/blob/master/oauth2/client-credentials-flow/token-generation/nginx.conf#L82)

File: src/client.coffee [hunk 2]
```diff
@@ -238,6 +238,45 @@ module.exports = class Client
...
+    
+    Parameters:
+      options is a Hash object with the following fields
+        service_token Required if you didn't use provider_key to ceate the Client instance
+        user_key Required
+        service_id Required (from November 2016)
+      callback {Function} Is the callback function that receives the Response object which includes `is_success` method to determine the status of the response
+
+    Example using provider_key:
+      client.authrep_with_user_key {service_id: '1234567890987', user_key: 'ca5c5a49'}, (response) ->
```
Comment (kevprice83) [inline src/client.coffee]: these examples should show `client_id` or `app_id`

File: src/client.coffee [hunk 3]
```diff
@@ -238,6 +238,45 @@ module.exports = class Client
...
+        if response.is_success
+          # All Ok
+        else
+         sys.puts "#{response.error_message} with code: #{response.error_code}"
+  
+    Example using service_token:
+      client.authrep_with_user_key {service_token: '12sdtsdr23454sdfsdf', service_id: '1234567890987', user_key: 'ca5c5a49'}, (response) ->
```
Comment (kevprice83) [inline src/client.coffee]: the same again here

**Comment by kevprice83** [comment]
@picsoung would you mind giving this a quick review or test for @avilatusell once it's ready?

**Comment by kevprice83** [comment]
@avilatusell what about the tests for the new method?

**Comment by picsoung** [comment]
Looks good @avilatusell :)

Agree with @kevprice83 comments.
+ If there was a test suite, that would be awesome

File: src/client.coffee [hunk 4]
```diff
@@ -238,6 +238,45 @@ module.exports = class Client
...
+        if response.is_success
+          # All Ok
+        else
+         sys.puts "#{response.error_message} with code: #{response.error_code}"
+  ###
+
+  authrep_with_oauth: (options, callback) ->
```
Comment (kevprice83) [inline src/client.coffee]: `oauth_authrep` would probably be more consistent naming with all the other functions.

File: test/client_test.coffee [hunk 5]
```diff
@@ -143,6 +146,35 @@ describe 'Basic test for the 3Scale::Client', ->
...
+        reqheaders:
+          'Host': 'su1.3scale.net'
+          'X-3scale-User-Agent': 'plugin-node-v' + require('../package.json').version
+
+      match = nock('https://su1.3scale.net', opts)
+        .get('/transactions/oauth_authorize.xml?service_id=1234567890987&app_id=foo&provider_key=1234abcd')
+        .reply(200, '<status><authorized>true</authorized><plan>Basic</plan></status>')
+
+      client = new Client '1234abcd'
+      client.oauth_authorize { service_id: '1234567890987', app_id: 'foo' }, (response) ->
```
Comment (kevprice83) [inline test/client_test.coffee]: This is calling the wrong method here. Should be `oauth_authrep`

File: test/client_test.coffee [hunk 6]
```diff
@@ -143,6 +146,35 @@ describe 'Basic test for the 3Scale::Client', ->
...
+      assert.throws (() -> client.authrep_with_oauth({}, () ->)), 'missing app_id'
+
+    it 'should include the Host and X-3scale-User-Agent headers', (done) ->
+      opts =
+        reqheaders:
+          'Host': 'su1.3scale.net'
+          'X-3scale-User-Agent': 'plugin-node-v' + require('../package.json').version
+
+      match = nock('https://su1.3scale.net', opts)
+        .get('/transactions/oauth_authorize.xml?service_id=1234567890987&app_id=foo&provider_key=1234abcd')
```
Comment (kevprice83) [inline test/client_test.coffee]: It should be calling _/oauth_authrep.xml_ here

File: test/client_test.coffee [hunk 7]
```diff
@@ -143,6 +146,35 @@ describe 'Basic test for the 3Scale::Client', ->
...
+  describe 'Request headers in authrep_with_oauth calls', ->
+    it 'should throw an exception if authrep_with_oauth called without :app_id', ->
+      client = new Client()
+      assert.throws (() -> client.authrep_with_oauth({}, () ->)), 'missing app_id'
+
+    it 'should include the Host and X-3scale-User-Agent headers', (done) ->
+      opts =
+        reqheaders:
+          'Host': 'su1.3scale.net'
+          'X-3scale-User-Agent': 'plugin-node-v' + require('../package.json').version
```
Comment (kevprice83) [inline test/client_test.coffee]: I think all OAuth calls should be sending the header `X-3scale-OAuth2-Grant-Type` but maybe @mpguerra can comment if that is critical or not as we are not specifically implementing one flow or another here.

**Review by avilatusell — COMMENTED** [review]
The tests for oauth_authrep fails. Need improvement.
"'Uncaught Error: Nock: No match for request GET https://su1.3scale.net/transactions/oauth_authrep.xml?service_id=1234567890987&app_id=ERROR&usage%5Bhits%5D=1&provider_key=1234abcd'"

@picsoung would you mind have a look?

**Comment by picsoung** [comment]
@avilatusell 
I've found the way to do the testing using `.query`

```
.get('/transactions/oauth_authrep.xml')
 .query({
   service_id: 1234567890987,
   app_id: "foo",
   "usage[hits]": 1,
    provider_key: "1234abcd"
 })
```

Once you modify it we can merge :D
good job!

**Review by avilatusell — COMMENTED** [review]
@picsoung 
Nice! thanks! :-)

**Comment by picsoung** [comment]
Awesome!
You can now do a "patch" version, merge the branch and publish it to NPM.

**Pull request merged**